src/acl_usm.cpp

```diff
@@ -396,6 +411,13 @@ clSharedMemAllocINTEL(cl_context context, cl_device_id device,
       }
       alloc_flags = *(properties + 1);
     } break;
+    case CL_MEM_ALLOC_BUFFER_LOCATION_INTEL: {
+      if (seen_flags.insert(CL_MEM_ALLOC_BUFFER_LOCATION_INTEL).second == false) {
+        UNLOCK_BAIL_INFO(CL_INVALID_PROPERTY, context,
+                         "Property specified multiple times");
+      }
+      mem_id = (cl_int) * (properties + 1);
```

Comment:
Should this validate the property value passed in from the (OpenCL) user and return an error?

Reply:
I was wondering about this as well.  The issue is that we may not have loaded the appropriate boardspec/autodiscovery string, so we actually have no way of knowing which buffer locations are valid.  I'm adding some functionality to the sim MMD to preload the boardspec, so I can do checking there.  I think in general it is reasonable to push this responsibility on to the MMD.  I'll make a note about that in `aocl_mmd.h`.

Reply:
Lower bound of 0 is a good point.  I changed it to cl_uint, so shouldn't need to worry about that any more.  As for the upper limit, I don't think there is any way to enforce this.  There may be a use-case for having arbitrary buffer locations?  Not sure.

Reply:
> I changed it to cl_uint, so shouldn't need to worry about that any more.

The [SYCL extension specifies a signed integer](https://github.com/intel/llvm/blob/db037ca70e1278bf5ee9ccb5c91d9e10ce7fb418/sycl/doc/extensions/proposed/sycl_ext_intel_runtime_buffer_location.asciidoc?plain=1#L146-L154) though, so we would either need to match that in the runtime, or change the SYCL extension which is still experimental.

@tiwaria1 What is the reason for using a signed integer for the buffer location property in the SYCL API?

Reply:
Oh really!? Interesting.  That is a bit odd.  Not sure why that would be.  I was just copying what we'd done in the device allocation code.  Should probably follow the spec if there is a good reason for it though!

Reply:
Pardon, the correct specification to consider is the OpenCL extension, [cl_intel_mem_alloc_buffer_location which specifies a `cl_uint`](https://github.com/KhronosGroup/OpenCL-Docs/blob/46090155cc32baa68057d5f58a5bc7a302af3b33/extensions/cl_intel_mem_alloc_buffer_location.asciidoc?plain=1#L135-L137), so your current code is good. @tiwaria1, what are your thoughts on modifying the SYCL extension specification to match the OpenCL counterpart?

Reply:
> Lower bound of 0 is a good point. I changed it to cl_uint, so shouldn't need to worry about that any more. As for the upper limit, I don't think there is any way to enforce this. There may be a use-case for having arbitrary buffer locations? Not sure.

For Full System compiles the valid buffer_location range is `[0, NUM_GLOBAL_MEMORIES-1]`, where `NUM_GLOBAL_MEMORIES` can be inferred from the FPGA `board_spec.xml` file.

For IPA, the range is a unique set of positive integers >= `0`, eg: it can be `[0,1,2,3]` or `[1024,1025,1026]` etc. (Since these are ids that the user is assigning to their memory interfaces). This information is present in the `ipinterfaces.xml` file that we generate. For 2022.3 we can enforce a limit asking users to use positive integers starting from 0 and incrementing by 1 if that is absolutely needed.